src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -317,6 +333,36 @@ public WorkspaceAnalyzerConfigOptions(AnalyzerConfigOptionsResult analyzerConfig
             }
         }
 
+        private sealed class WorkspaceSyntaxTreeOptionsProvider : SyntaxTreeOptionsProvider
+        {
+            private readonly ValueSource<AnalyzerConfigSet> _lazyAnalyzerConfigSet;
+
+            public WorkspaceSyntaxTreeOptionsProvider(ValueSource<AnalyzerConfigSet> lazyAnalyzerConfigSet)
+                => _lazyAnalyzerConfigSet = lazyAnalyzerConfigSet;
+
+            public override bool? IsGenerated(SyntaxTree tree)
+            {
+                var options = _lazyAnalyzerConfigSet
+                    .GetValue(CancellationToken.None).GetOptionsForSourcePath(tree.FilePath);
+                return GeneratedCodeUtilities.GetIsGeneratedCodeFromOptions(options.AnalyzerOptions);
+            }
+
+            public override bool TryGetDiagnosticValue(SyntaxTree tree, string diagnosticId, out ReportDiagnostic severity)
+            {
+                var options = _lazyAnalyzerConfigSet
+                    .GetValue(CancellationToken.None).GetOptionsForSourcePath(tree.FilePath);
```

Comment:
Will this re-parse the editorconfig and generate a new `AnalyzerConfigOptionsResult` every time? Is there some caching that can be done if the associated editorconfig is the same for the entire project/solution?

Reply:
The _lazyAnalyzerConfigSet is the AsyncLazy that will cache the _parsed_ .editorconfigs, so we won't be reparsing them. This however isn't caching the result that applies to a SyntaxTree, we'll recompute that bit each time. Whether that's good or not it's hard to say...

Reply:
For now we'll go this way -- if we want to add an extra cache atop then we can do so if we see any issues. The cache would be trivial to add.